Description: Work around php5-mongo's bugs
 PHP's MongoDB driver incorrectly disposes of the variable it does
 not own. This patch tries to prevent memory corruption.
Author: Vladimir Kolesnikov <vladimir@extrememember.com>
Origin: https://github.com/phalcon/cphalcon/issues/851
Bug: https://github.com/phalcon/cphalcon/issues/851
Forwarded: https://github.com/phalcon/cphalcon/issues/852
Last-Update: 2013-06-16

--- php-phalcon-1.2.0.orig/cphalcon/ext/mvc/collection.c
+++ php-phalcon-1.2.0/cphalcon/ext/mvc/collection.c
@@ -1424,7 +1424,17 @@ PHP_METHOD(Phalcon_Mvc_Collection, save)
 	 * Save the document
 	 */
 	PHALCON_INIT_NVAR(status);
+#if PHP_VERSION_ID < 50400
+	{
+		zval *params[2] = { data, options };
+		zval func;
+		INIT_ZVAL(func);
+		ZVAL_STRING(&func, "save", 0);
+		call_user_function(EG(function_table), &collection, &func, status, 2, params TSRMLS_CC);
+	}
+#else
 	phalcon_call_method_p2(status, collection, "save", data, options);
+#endif
 	if (Z_TYPE_P(status) == IS_ARRAY) { 
 		if (phalcon_array_isset_string(status, SS("ok"))) {
 	
--- php-phalcon-1.2.0.orig/cphalcon/build/64bits/phalcon.c
+++ php-phalcon-1.2.0/cphalcon/build/64bits/phalcon.c
@@ -47288,7 +47288,18 @@ static PHP_METHOD(Phalcon_Mvc_Collection
 	add_assoc_bool_ex(options, SS("safe"), 1);
 	
 	PHALCON_INIT_NVAR(status);
+#if PHP_VERSION_ID < 50400
+	{
+		zval *params[2] = { data, options };
+
+		zval func;
+		INIT_ZVAL(func);
+		ZVAL_STRING(&func, "save", 0);
+		call_user_function(EG(function_table), &collection, &func, status, 2, params TSRMLS_CC);
+	}
+#else
 	phalcon_call_method_p2_key(status, collection, "save", data, options, 210727548372UL);
+#endif
 	if (Z_TYPE_P(status) == IS_ARRAY) { 
 		if (phalcon_array_isset_quick_string(status, SS("ok"), 193501407UL)) {
 	
--- php-phalcon-1.2.0.orig/cphalcon/build/32bits/phalcon.c
+++ php-phalcon-1.2.0/cphalcon/build/32bits/phalcon.c
@@ -52035,7 +52035,18 @@ static PHP_METHOD(Phalcon_Mvc_Collection
 	add_assoc_bool_ex(options, SS("safe"), 1);
 	
 	PHALCON_INIT_NVAR(status);
+#if PHP_VERSION_ID < 50400
+	{
+		zval *params[2] = { data, options };
+
+		zval func;
+		INIT_ZVAL(func);
+		ZVAL_STRING(&func, "save", 0);
+		call_user_function(EG(function_table), &collection, &func, status, 2, params TSRMLS_CC);
+	}
+#else
 	phalcon_call_method_p2_key(status, collection, "save", data, options, 274150868UL);
+#endif
 	if (Z_TYPE_P(status) == IS_ARRAY) { 
 		if (phalcon_array_isset_quick_string(status, SS("ok"), 193501407UL)) {
 	
--- php-phalcon-1.2.0.orig/cphalcon/build/safe/phalcon.c
+++ php-phalcon-1.2.0/cphalcon/build/safe/phalcon.c
@@ -52035,7 +52035,18 @@ static PHP_METHOD(Phalcon_Mvc_Collection
 	add_assoc_bool_ex(options, SS("safe"), 1);
 	
 	PHALCON_INIT_NVAR(status);
+#if PHP_VERSION_ID < 50400
+	{
+		zval *params[2] = { data, options };
+
+		zval func;
+		INIT_ZVAL(func);
+		ZVAL_STRING(&func, "save", 0);
+		call_user_function(EG(function_table), &collection, &func, status, 2, params TSRMLS_CC);
+	}
+#else
 	phalcon_call_method_p2(status, collection, "save", data, options);
+#endif
 	if (Z_TYPE_P(status) == IS_ARRAY) { 
 		if (phalcon_array_isset_string(status, SS("ok"))) {
 	
